### ISSUE TITLE
Python: TSNE().fit_transform endless loop with CUDA 9.2

### ISSUE THREAD

**Issue by rappdw** [CannyLab/tsne-cuda#14]
I'm interested in looking at this implementation. To facilitate my exploration, I've created a docker image and am attempting to run this on an EC2 p3.xLarge instance (with [nvidia-docker2](https://github.com/NVIDIA/nvidia-docker) installed.) ([Dockerfile](https://github.com/CannyLab/tsne-cuda/files/2310812/Dockerfile.txt) is attached as is the [script](https://github.com/CannyLab/tsne-cuda/files/2310813/install-mkl.sh.txt) to install mkl in the docker image)

I had to modify the CMakeLists.txt to support the Nvidia Tesla v100 available on the p3 instance by adding the following to CMakeLists.txt:

`63:                     -gencode=arch=compute_70,code=sm_70 `

After building and running the docker file, I try the simple example for python on the wiki:
```
>>> import numpy as np
>>> from tsnecuda import TSNE
>>> X = np.array([[0, 0, 0], [0, 1, 1], [1, 0, 1], [1, 1, 1]])
>>> X_embedded = TSNE(perplexity=64.0, learning_rate=270).fit_transform(X)
WARNING clustering 4 points to 2 centroids: please provide at least 78 training points
```
fit_transform never returns a result. One of the CPU cores on the p3 is running python at 100% and nvidia-smi shows GPU utilization at 100%.

To reproduce this:
1) Start a p3.xLarge EC2 instance
2) install nvidia-docker
3) clone repo and modify CMakeLists.txt as described
4) save attached Dockerfile and install-mkl.sh to <repo>/docker/test/
5) from repo directory:
  * `docker build -t tsne-cuda-test:latest -f docker/test/Dockerfile .`
  * `docker run --init -it --runtime=nvidia -e NVIDIA_VISIBLE_DEVICES=all tsne-cuda-test:latest`
6) start Python REPL and entry example above

**Comment by DavidMChan** [comment]
Hmm, So I am currently not able to reproduce this bug using a GPU with compute capability 5.2, and CUDA 9.0. (Using the provided steps). So I think that it's an interesting interaction with the p3.xLarge EC2 setup. Can you run with "verbose=1", and report the results?

**Comment by rappdw** [comment]
>>> X_embedded = TSNE(perplexity=64.0, learning_rate=270, verbose=1).fit_transform(X)
WARNING clustering 4 points to 2 centroids: please provide at least 78 training points
[Step 0] Avg. Gradient Norm: 0.0143171

**Comment by rappdw** [comment]
@DavidMChan Any further thoughts on this? (I posted the only output received when enabling verbose just above.)

**Comment by DavidMChan** [comment]
Sorry - It's been a busy start to the semester. It's a bit tricky to debug without powering up an EC2 instance (since nobody here had a V100 gpu). Can you report if changing the perplexity or learning rate changes the outcome? We're tracking a bug with the launch parameters which can cause it to hang on certain GPUs.

**Comment by rappdw** [comment]
Tried perplexity @ 50.0 +/- 25.0 and learning_rate @ 200.0 +/- 100.0. Same results as before (gradient norm @ step 0 varied, but program still caught in loop)

**Comment by rmrao** [comment]
Do you know if this happens on all inputs X or just on this input? So if you set `X = np.random.random((5000, 50))` does this still hang?

I ask because I just added a bug that we've been seeing for a while (#17) that may be the culprit. If this is a data-dependent bug and you're able to run on the random input, then it's probably the same issue. However, if you're seeing this on all input data then it's probably something different.

**Comment by rappdw** [comment]
X = np.random.rand(10000, 150)

causes a hang

**Comment by rmrao** [comment]
Huh. My next thought was that something could be wrong with the launch bounds if its failing on all inputs, but since it is printing out the average gradient norm at step 0, that means it's not instantly failing. Which means that it probably *is* the same issue as #17.

This is definitely an issue we are trying to solve, however we're all fairly busy with other projects at the moment, so it may be some time before it gets fixed. In the meantime, since we seem to experience the issue less frequently with our setup, you may have better luck with a lower CUDA version (9.0 or 8.0) or a different GPU (most of our testing was done on a Titan Xp), if either of those are an option for you.

@DavidMChan thoughts?

**Comment by rappdw** [comment]
Some additional data...

The build doesn't appear to be stable. If I generate the makefiles and then run make, generate a wheel from that and install it, the code hangs as described above. If I run make a second time, it recompiles about half of the project. If I run it a third time, it recompiles one file. If I run it a fourth time, nothing is recompiled. If I use the wheel generated from this final state, the code works fine.

**Comment by rmrao** [comment]
Yes, there's definitely some dependency resolution issue with parallelizing `make`. We don't really know what the issue is or how to go about fixing it, it may be an issue with `nvcc`'s detection of dependencies. We typically solve this by running `make -jN` and finishing with `make -j1`.

**Comment by rmrao** [comment]
Also we've made some progress on the issue of the code hanging on all inputs. It looks like this is a separate issue from #17. We were able to duplicate it on a Titan Xp while trying to do the conda build.

It seems like the most likely cause is incorrect launch bounds. The launch bounds are defined in [`src/include/options.h`](https://github.com/CannyLab/tsne-cuda/blob/1249948704b0ae1847ebe614801f8a326050b0f4/src/include/options.h#L204-L219). It looks like the bounds for `architecture >= 6` may be wrong. We found that using the bounds we define for `architecture == 5` and compiling from source seems to work.

We haven't made this into an update yet because we're not completely certain that it is the issue. For one, our conda build is still experiencing the hanging issue even with these changes. It's unclear whether that's because we haven't isolated the source of the problem or because we're doing something wrong on the conda build portion.

If you'd like to try the fix, I suggest editing the launch bounds to remove the `architecture >= 6` section and having it default to the `architecture == 5` section. Otherwise we will hopefully nail down the issue in the next week or two and make an update then.

**Comment by rappdw** [comment]
Also, it seems like an `architecture == 7` logic branch on the launch bounds in [src/include/options.h](https://github.com/CannyLab/tsne-cuda/blob/1249948704b0ae1847ebe614801f8a326050b0f4/src/include/options.h#L204-L219) would be in order and I can give it a shot if you provide appropriate values...

**Comment by rmrao** [comment]
Just completed a major refactor in #39 and pushed that to a [new release](https://github.com/CannyLab/tsne-cuda/releases/tag/2.0). The changes massively improve the stability of the algorithm, fix the multi-threaded compilation issue, add support for CUDA 10 and compute architecture 7, and also decrease runtime and memory usage.

If you see any of these problems occurring with the new release, please let us know in a new issue.

**Comment by CeyaoZhang** [comment]
I encountered the same problem, orz!

**Comment by DavidMChan** [comment]
@CeyaoZhang  - Can you please open a new issue on the repository with your system details?

**Comment by CeyaoZhang** [comment]
Thanks, I solved this problem.
And I don't think this is due to the system configuration, it's just that the input data/samples are fewer than the requirement.

In the https://github.com/CannyLab/tsne-cuda/issues/14, the author provides only 4 points and gets a warning.
> WARNING clustering 4 points to 2 centroids: please provide at least 78 training points

When I provide more data, the problem is solved. For example,

```
import numpy as np
from tsnecuda import TSNE

X = np.random.randn(2000, 3)
X_embedded = TSNE(perplexity=64.0, learning_rate=270).fit_transform(X)
X_embedded.shape
```

Thanks again for your fantastic work, **tsne-cuda** is really faster than the others.
